### PR TITLE
SSL support for MQTT connection

### PR DESCRIPTION
**Pull request by kleini**
This is stolen from timpur's v2.1 branch.
I like to encrypt my MQTT connections because passwords are transmitted over MQTT in clear text. Passwords transmitted in clear text should be always avoided.

### CODE REVIEW

**Comment by timpur** [comment]
Just double check my change for the Mac string 

https://github.com/homieiot/homie-esp8266/blob/48e8fab8e69744f33e56355233a7a028d9294601/src/Homie/Limits.hpp#L13

:P

**Comment by amayii0** [comment]
Just a very generic comment, I'm not following heavily the dev but looking here and there.
Would be great if consts are commented with a sample value, from this, how are you storing MAC?

**Comment by kleini** [comment]
Up to this point I only picked some commits. Will read and try to verify the code now.

**Comment by kleini** [comment]
@timpur I hope I fixed all char array and string length.

**Comment by timpur** [comment]
Looks good (I believe these changes were in 2.1)

**Comment by kleini** [comment]
Just to mention it, because I stumbled again over it:
ASYNC_TCP_SSL_ENABLED requires PIO_FRAMEWORK_ARDUINO_LWIP2_HIGHER_BANDWIDTH otherwise the SSL connections do not work at all.

**Comment by timpur** [comment]
> Just to mention it, because I stumbled again over it:
> ASYNC_TCP_SSL_ENABLED requires PIO_FRAMEWORK_ARDUINO_LWIP2_HIGHER_BANDWIDTH otherwise the SSL connections do not work at all.

Maybe @kleini document that ?

**Comment by kleini** [comment]
@timpur documented SSL connections for MQTT.

File: docs/configuration/json-configuration-file.md [hunk 1]
```diff
@@ -46,6 +48,7 @@ Here are the rules:
    - `bssid` and `channel` have to be defined together and these settings are independand of settings related to static IP
    - to define static IP, `ip` (IP address), `mask` (netmask) and `gw` (gateway) settings have to be defined at the same time
    - to define second DNS `dns2` the first one `dns1` has to be defined. Set DNS without `ip`, `mask` and `gw` does not affect the configuration (dns server will be provided by DHCP). It is not required to set DNS servers.
+* `ssl_fingerprint` is not required if `ssl` is enabled.
```
Comment (bodiroga) [inline docs/configuration/json-configuration-file.md]: Is this sentence correct? If 'ssl' is disabled, where is 'ssl_fingerprint' used? I don't get it... 😕

**Comment by bodiroga** [comment]
Should we merge this PR in the 'develop' branch or the 'develop-v3' branch? IMHO, we should leave 'develop' only for bugfixes and add new features to 'develop-v3' 😉

File: docs/configuration/json-configuration-file.md (hunk 1 above)
Comment (kleini) [reply docs/configuration/json-configuration-file.md]: Okay, that sentence is weird. It is a result of editing and not reading again. Will fix this.

**Comment by kleini** [comment]
Yes, will rebase my work on develop-v3 the next days.

**Comment by stritti** [comment]
> Yes, will rebase my work on develop-v3 the next days.

@kleini could you rebase it to be merged then?

**Comment by kleini** [comment]
I rebased my work on develop-v3. The build is broken because of the resulting image size. Will try to check, why the image size is limited this way.

File: docs/configuration/json-configuration-file.md (hunk 1 above)
Comment (bodiroga) [reply docs/configuration/json-configuration-file.md]: `ssl_fingerprint` is required if `ssl` is enabled. :+1:

Comment (kleini) [reply docs/configuration/json-configuration-file.md]: Oh, I missed to fix that sentence. And ssl_fingerprint is optional.

Comment (bodiroga) [reply docs/configuration/json-configuration-file.md]: Oh, so even if 'ssl' is enabled, 'ssl_fingerprint' is not required? Then the sentence is correct, but we should rewrite it to make it cleanerer. Could you please commit a change? Thanks for your work @kleini!

Comment (kleini) [reply docs/configuration/json-configuration-file.md]: Fixed. I hope the text is more clear now.

**Comment by kleini** [comment]
@bodiroga This is not my work. @timpur implemented this and I just try to get it merged as I consider encryption for MQTT connections as essential (clear text passwords).

**Pull request merged**